sdk/dotnet/Pulumi/Serialization/ResourcePackages.cs

```diff
@@ -1,40 +1,98 @@
+        string Name { get; }
+        string? Version { get; }
+        ProviderResource ConstructProvider(string name, string type, string urn);
+        Resource Construct(string name, string type, string urn);
     }
 
     internal static class ResourcePackages
     {
-        internal static ConcurrentDictionary<string, IResourcePackage> _resourcePackages = new ConcurrentDictionary<string, IResourcePackage>();
+        private static ImmutableList<IResourcePackage>? _resourcePackages;
+        private static readonly object _resourcePackagesLock = new object();
 
-        private static string PackageKey(string name, string version)
+        internal static bool TryGetResourcePackage(string name, string? version, [NotNullWhen(true)] out IResourcePackage? package)
         {
-            return $"{name}@{version}";
+            lock (_resourcePackagesLock)
+            {
+                _resourcePackages ??= DiscoverResourcePackages();
+            }
+
+            var minimalVersion = version != null ? SemVersion.Parse(version) : new SemVersion(0);
+            var matches =
+                from p in _resourcePackages
+                where p.Name == name
+                let packageVersion = p.Version != null ? SemVersion.Parse(p.Version) : minimalVersion
+                where packageVersion >= minimalVersion
+                where (version == null || p.Version == null || minimalVersion.Major == packageVersion.Major)
+                orderby packageVersion descending
+                select p;
+            
+            package = matches.FirstOrDefault();
+            return package != null;
         }
 
-        internal static bool TryGetResourcePackage(string name, string version, [NotNullWhen(true)] out IResourcePackage? package)
+        private static ImmutableList<IResourcePackage> DiscoverResourcePackages()
         {
-            return _resourcePackages.TryGetValue(PackageKey(name, version), out package);
+            return LoadReferencedAssemblies()
+                .SelectMany(s => s.GetTypes())
+                .Where(typeof(IResourcePackage).IsAssignableFrom)
+                .SelectMany(type => type.GetConstructors().Where(c => c.GetParameters().Length == 0))
+                .Select(c => c.Invoke(new object[0]))
+                .Cast<IResourcePackage>()
+                .ToImmutableList();
         }
-
-        public static void RegisterResourcePackage(string name, string version, IResourcePackage package)
+        
+        // Assemblies are loaded on demand, so it could be that some assemblies aren't yet loaded to the current
+        // app domain at the time of discovery. This method iterates through the list of referenced assemblies
+        // recursively.
+        // Note: If an assembly is referenced but its types are never used anywhere in the program, that reference
+        // will be optimized out and won't appear in the result of the enumeration.
```

Comment:
How would someone mitigate this in their program if they're running into it?

Reply:
They would have to use something from that assembly in their code. This won't happen often in practice, I suspect.